### PR TITLE
Refactor syncing and clearing of patient data out of UserSession

### PR DESCRIPTION
**Pull request by vinaysshenoy**
https://www.pivotaltracker.com/story/show/168980786

### CODE REVIEW

File: app/src/main/java/org/simple/clinic/forgotpin/confirmpin/ForgotPinConfirmPinScreenController.kt [hunk 1]
```diff
@@ -102,7 +101,7 @@ class ForgotPinConfirmPinScreenController @Inject constructor(
         .share()
 
     val makeResetPinCall = validPin
-        .flatMapSingle { newPin -> syncAndClearPatientData().toSingleDefault(newPin) }
+        .flatMapSingle { newPin -> syncAndClearPatientData.run().toSingleDefault(newPin) }
```
Comment (ragunathjawahar) [inline app/src/main/java/org/simple/clinic/forgotpin/confirmpin/ForgotPinConfirmPinScreenController.kt]: Curious, why do we have to clear patient data while resetting the user's pin?

File: app/src/test/java/org/simple/clinic/user/clearpatientdata/SyncAndClearPatientDataTest.kt [hunk 2]
```diff
@@ -34,20 +34,6 @@ class SyncAndClearPatientDataTest {
   private val prescriptionPullToken = mock<Preference<Optional<String>>>()
   private val medicalHistoryPullToken = mock<Preference<Optional<String>>>()
 
-  private val syncAndClearPatientData = SyncAndClearPatientData(
```
Comment (ragunathjawahar) [inline app/src/test/java/org/simple/clinic/user/clearpatientdata/SyncAndClearPatientDataTest.kt]: Curious, why are we doing this?

File: app/src/main/java/org/simple/clinic/login/LoginUserWithOtp.kt [hunk 3]
```diff
@@ -22,7 +22,7 @@ import javax.inject.Named
 
 class LoginUserWithOtp @Inject constructor(
     private val loginApi: LoginApi,
-    private val dataSync: Lazy<DataSync>,
```
Comment (ragunathjawahar) [inline app/src/main/java/org/simple/clinic/login/LoginUserWithOtp.kt]: 🚀

File: app/src/main/java/org/simple/clinic/forgotpin/confirmpin/ForgotPinConfirmPinScreenController.kt (hunk 1 above)
Comment (vinaysshenoy) [reply app/src/main/java/org/simple/clinic/forgotpin/confirmpin/ForgotPinConfirmPinScreenController.kt]: This is a product decision. We don't know if the user who is attempting to reset the PIN is a legitimate user who has genuinely forgotten their PIN, or someone who is trying to reset the PIN to get access to the data on the phone. To be safe, we clear the patient data stored locally when a user tries to reset their a PIN.

This is expected to be paired with a phone call from the user's supervisor to verify that they are the real user who is trying to reset the PIN before they get access to the patient data again.

File: app/src/test/java/org/simple/clinic/user/clearpatientdata/SyncAndClearPatientDataTest.kt (hunk 2 above)
Comment (vinaysshenoy) [reply app/src/test/java/org/simple/clinic/user/clearpatientdata/SyncAndClearPatientDataTest.kt]: In a later commit, I need to override one of the properties passed to the constructor. So I made this change.

**Pull request merged**